### DISCUSSION TITLE
createModule

### DISCUSSION THREAD

**Discussion by Kelin2025** [zerobias/effector#23] [Archive]
Hi. I want to propose a helper to create database-like stores where each field is a store and each record has its ID. 

### Automatic ID generation by entity name
```js
const todos = createModule({
  entity: 'todo'
})

todos.add()
todos.add()

todos.entities.getState()
/*
  { 
    'todo_1': { _edited: null, _relatedTo: null, 
    'todo_2': { _edited: null, _relatedTo: null 
  }
*/
```

### Automatic fields creation
```js
const todos = createModule({
  entity: 'todo',
  fields: ['text', 'isActive']
})

todos.add({
  text: 'Use effector',
  isActive: true
})

todos.entities.getState()
// { 'todo_1': { _edited: null, _relatedTo: null, text: 'Use effector', isActive : true } }
```
> I think better use Vue props-like notation `{ text: { type: String, default: '' } }` for fields. It also would be better for typings

### Fields are accessible as well
```js
todos.fields.text.getState()
// { 'todo_1': 'Use effector' }
```

### Items removal
```js
todos.removeMany(['todo_1', 'todo_2'])
todos.removeOne('todo_1')
todos.removeByRelation('parent_1')
```
> Cleanup of stores is synced so records will be removed from all fields stores

### Edit feature
It's OK when you have todo form with only one field, you can create store and its logic but if you have complex entities and editors you might get a lot of code duplication. To fix that, you can create a temporary clone of records which will replace original then.
```
// Will create todo with id 'todo_1' and _edited 'todo_1'
todos.create({ text: 'Hello' })

// Will remove _edited from todo with id 'todo_1'
todos.save('todo_1')

// Will create todo with id 'todo_2' and _edited 'todo_1'
todos.edit('todo_1')

// Will replace 'todo_1' with 'todo_2' but with id 'todo_1'. 'todo_2' record will be removed
todos.save('todo_2')

// Computed store to get only saved entities
todos.saved.getState()
```

### WIP 
- Add relations feature to get parents and children + sync add/remove of them
- Write typings - I'm not a typings guru so I can fail that

[**Go to sandbox**](https://codesandbox.io/s/0pk9jy625p)

**Comment by zerobias** [comment]
Hi, I'm going to reuse api for model declaration from here https://github.com/zerobias/relativism 
Is it possible to combine them together?

**Comment by sergeysova** [comment]
@Kelin2025 	What about publish that example as side package?

**Reply by sergeysova** [reply]
@Kelin2025 Any updates here?

**Reply by Kelin2025** [reply]
Meh, I feel like it's not the way to go, at least the API should be completely different
